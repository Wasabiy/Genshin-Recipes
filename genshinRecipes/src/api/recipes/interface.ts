export interface Food {
    name?: string,
    rarity?: number,
    type?: Type,
    effect?: string,
    hasRecipe?: boolean,
    description?: string,
    character?: string,
    baseDish?: string,
    recipe?: Recipe[],
    proficiency?: number | string,
    event?: string,
    charcter?: string,
    food: any
}

export interface Recipe {
    item:     string;
    quantity: number;
}

export enum Type {
    ATKBoostingDish = "ATK-Boosting Dish",
    AdventurerSDish = "Adventurer's Dish",
    DEFBoostingDish = "DEF-Boosting Dish",
    RecoveryDish = "Recovery Dish",
}