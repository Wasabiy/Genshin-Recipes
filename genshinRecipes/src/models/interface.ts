export type KeyFood = {
    key: string,
} & Food

export interface Food {
    name: string,
    rarity: number,
    type: DishType,
    effect: string,
    hasRecipe?: boolean,
    description: string,
    character?: string,
    baseDish?: string,
    recipe?: Recipe[],
    proficiency: number | string,
    event?: string,
    isFavorited?: boolean,
}

export interface Recipe {
    item:     string;
    quantity: number;
}
export interface Ingredient{
    name: string,
    description: string,
    sources: string[]
}
export type KeyIngredient = {
    key: string,
} & Ingredient

export enum DishType {
    ATKBoostingDish = "ATK-Boosting Dish",
    AdventurerSDish = "Adventurer's Dish",
    DEFBoostingDish = "DEF-Boosting Dish",
    RecoveryDish = "Recovery Dish",
}