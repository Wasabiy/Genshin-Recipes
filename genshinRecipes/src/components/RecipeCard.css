.cardSquares {
  display: flex;
  flex-direction: column;
  background-color: none;
  margin: 3px;
  padding: 5px;
  width: 120px;
  height: 120px;
  font-size-adjust: inherit;
}
.imageButton {
  position: relative;
  margin: 0;
}

.likedButton {
  position: absolute;
  border: none;
  height: 2rem;
  top: 2px;
  right: 2px;
}

.normalButton {
  position: absolute;
  border: none;
  height: 2rem;
  top: 2px;
  right: 2px;
  border-radius: 10px;
}

.normalButton:hover {
  background-color: rosybrown;
}

.likedButton:hover {
  background-color: rosybrown;
  border-radius: 10px;
}

.recipeImages {
  display: flex;
  width: 80px;
  height: 80px;
  margin: 0;
  padding: 0;
}
.recipeNames {
  display: flex;
  margin: 2px;
  font-size: inherit;
}
