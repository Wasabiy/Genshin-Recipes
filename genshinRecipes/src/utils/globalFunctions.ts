//reusable function

import { KeyFood } from '../models/interface.ts';

// @ts-expect-error any type error
export function reformatData(data, type) {
  console.log(type);
  return Object.entries(data).map((value: [string, any]) => {
    const key = value[0];
    const mat = value[1];
    return {
      ...mat,
      key: key,
    } as typeof type;
  });
}

export function getFavoriteState(food: KeyFood) {
  try {
    const localIsLiked = localStorage.getItem(food.key);
    if (localIsLiked != null) {
      return true;
    }
  } catch (e) {
    console.log(e);
    return false;
  }
}

export function changeFavoriteState(food: KeyFood) {
  const localIsLiked = localStorage.getItem(food.key);
  if (localIsLiked == null) {
    localStorage.setItem(food.key, JSON.stringify({ ...food }));
  } else {
    localStorage.removeItem(food.key);
  }
}
