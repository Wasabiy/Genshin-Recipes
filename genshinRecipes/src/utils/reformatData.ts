//gjenbrukbar funksjon

export function reformatData(data,type) {
    return Object.entries(data).map((value: [string, any]) => {
        const key = value[0];
        const mat = value[1];
        return {
            ...mat,
            key: key,
        } as typeof type;
    });
};