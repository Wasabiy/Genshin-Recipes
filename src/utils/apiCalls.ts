export const fetchFood = async () => {
  const res = await fetch('https://genshin.jmp.blue/consumables/food/', {
    method: 'GET',
    mode: 'cors',
    headers: {
      'Access-Control-Allow-Origin': 'https://genshin.jmp.blue/materials/cooking-ingredients/',
    },
  });
  return res.json();
};

export const fetchIngredient = async () => {
  const res = await fetch('https://genshin.jmp.blue/materials/cooking-ingredients/', {
    method: 'GET',
    mode: 'cors',
    headers: {
      'Access-Control-Allow-Origin': 'https://genshin.jmp.blue/consumables/food/',
    },
  });
  return res.json();
};
