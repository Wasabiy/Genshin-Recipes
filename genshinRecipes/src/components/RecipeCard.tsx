import { useEffect, useState } from 'react';
import '../style/components/RecipeCard.css';
import 'react-router-dom';
import { KeyFood } from '../models/interface.ts';
import {  useNavigate } from 'react-router-dom';
import { changeFavoriteState, getFavoriteState } from '../utils/globalFunctions.ts';
import like from '../assets/like.png';
import liked from '../assets/liked.png';

function RecipeCard({ food, src, onFavoriteChanged }: { food: KeyFood; src: string; onFavoriteChanged?: () => void }) {
  const [isLiked, setIsLiked] = useState(Boolean);
  const navigate = useNavigate();
  useEffect(() => {
    if (getFavoriteState(food)) {
      setIsLiked(true);
    } else {
      setIsLiked(false);
    }
  }, [food]);

  const handleReadMore = () => {
    navigate('/recipeInfo', { state: { food: food } });
  };
  const handleFavoriteChange = () => {
    const likedBool = getFavoriteState(food);
    setIsLiked(!likedBool);
    changeFavoriteState(food);
    if (onFavoriteChanged) {
      onFavoriteChanged();
    }
  };

  return (
    <>

        <figure key={food.key} className="cardSquares">
          <figure className="imageButton">
            <img className="recipeImages" src={src} alt={food.name} width="50" height="60" onClick={handleReadMore}></img>
            <img
              alt={'likedButton'}
              src={isLiked ? liked : like}
              key={food.key + 'Btn'}
              onClick={(e) => {
                e.preventDefault();
                handleFavoriteChange();
              }}
              className={isLiked ? 'likedButton' : 'normalButton'}
            ></img>
          </figure>
          <p className="recipeNames">{food.name}</p>
        </figure>
    </>
  );
}

export default RecipeCard;
