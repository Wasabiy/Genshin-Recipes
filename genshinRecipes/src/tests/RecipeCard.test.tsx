import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';
import { describe, expect, test, beforeAll, vi } from "vitest";
import { server } from "./mock/mockServer.ts"
import { KeyFood, Type } from "../models/interface";
import RecipeCard from "../components/RecipeCard";
import { changeFavoriteState, getFavoriteState } from "../utils/globalFunctions";
import { MemoryRouter } from "react-router-dom";

vi.mock("../utils/globalFunctions", () => ({
    changeFavoriteState: vi.fn(),
    getFavoriteState: vi.fn(),
}));

const mockPush = vi.fn();

vi.mock("@hooks/useNavigation", () => {
  return {
    default: () => ({
      push: mockPush,
    }),
  };
});

const mockFood: KeyFood = {
    key: "sweet-madame",
    name: "Sweet Madame",
    rarity: 2,
    type: Type.RecoveryDish,
    effect: "Restores 20~24% of Max HP and an additional 900~1,500 HP to the selected character.",
    hasRecipe: true,
    description: "Honey-roasted fowl. The honey and sweet flowers come together to compliment the tender fowl meat.",
    proficiency: 10,
    recipe: [],
  };

const mockSrc:string = `https://genshin.jmp.blue/consumables/food/${mockFood.key}`;

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("RecipeCard.tsx snapshot test", () => {
    test('matches snapshot', () => {
        const { asFragment } = render(
            <MemoryRouter>
                <RecipeCard food={mockFood} src={mockSrc} />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('testing RecipeCard functionalities', () => {

        test('Check if isLiked state changes' , () => {
        const { getByRole } = render(
            <MemoryRouter>
                <RecipeCard food={mockFood} src={mockSrc} />
            </MemoryRouter>
        );

        const button = getByRole('button');

        fireEvent.click(button);
        expect(button).toHaveClass('likedButton');
        expect(changeFavoriteState).toHaveBeenCalledWith(mockFood);

        getFavoriteState.mockReturnValueOnce(true);
        fireEvent.click(button);
        expect(button).toHaveClass('normalButton');
        expect(changeFavoriteState).toHaveBeenCalledTimes(2);
    })

       test('Check if clicking on a card leads to RecipeInfoPage.tsx', async () => {
            render(<RecipeCard food={mockFood} src={mockSrc} />, {
                    wrapper: ({children}) => (
                      <MemoryRouter initialEntries={["/"]}>
                        {children}
                      </MemoryRouter>
                    ),
                  }
            );

            const cardSquares = (screen.getByRole('link', { name: /sweet madame/i }));
            fireEvent.click(cardSquares);
            expect(screen.getByText('Sweet Madame')).toBeTruthy();
        })

});