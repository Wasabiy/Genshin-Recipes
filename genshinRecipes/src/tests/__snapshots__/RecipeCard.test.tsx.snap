// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecipeCard.tsx snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <a
    class="infoLink"
    href="/recipes/sweet-madame"
  >
    <figure
      class="cardSquares"
    >
      <figure
        class="imageButton"
      >
        <img
          alt="Sweet Madame"
          class="recipeImages"
          height="60"
          src="https://genshin.jmp.blue/consumables/food/sweet-madame"
          width="50"
        />
        <img
          alt="likedButton"
          class="normalButton"
          src="/project1/src/assets/like.png"
        />
      </figure>
      <p
        class="recipeNames"
      >
        Sweet Madame
      </p>
    </figure>
  </a>
</DocumentFragment>
`;
